site/app/models/gradeable/Component.php

```diff
@@ -0,0 +1,274 @@
+    protected $ta_comment = "";
+    /** @property @var string The comment visible to the student */
+    protected $student_comment = "";
+    /** @property @var int The minimum points this component can contribute to the score (can be negative) */
+    protected $lower_clamp = 0;
+    /** @property @var int The number of points this component is worth with no marks */
+    protected $default = 0;
+    /** @property @var int The full value of this component without extra credit */
+    protected $max_value = 0;
+    /** @property @var int The maximum number of points this component can contribute to the score (can be > $max_value) */
+    protected $upper_clamp = 0;
+    /** @property @var bool If this is a text component (true) or a numeric component (false) for numeric/text components */
+    protected $text = false;
+    /** @property @var bool If this is a peer grading component */
+    protected $peer = false;
+    /** @property @var int The order of this component in the gradeable */
+    protected $order = -1;
+    /** @property @var int The pdf page this component will reside in */
+    protected $page = -1;
+
+    /** @property @var Mark[] All possible common marks that can be assigned to this component */
+    protected $marks = array();
+
+
+    public function __construct(Core $core, Gradeable $gradeable, $details, array $marks)
+    {
+        parent::__construct($core);
+
+        $this->setGradeable($gradeable);
+        $this->setMarks($marks);
+        $this->setIdInternal($details['id']);
+        $this->setTitle($details['title']);
+        $this->setTaComment($details['ta_comment']);
+        $this->setStudentComment($details['student_comment']);
+        $this->setPoints($details['lower_clamp'], $details['default'], $details['max_value'], $details['upper_clamp']);
+        $this->setText($details['text']);
+        $this->setPeer($details['peer']);
+        $this->setOrder($details['order']);
+        $this->setPage($details['page']);
+    }
+
+    /**
+     * Gets the component's gradeable
+     * @return Gradeable The component's gradeable
+     */
+    public function getGradeable()
+    {
+        return $this->gradeable;
+    }
+
+    /* Overridden setters with validation */
+    /**
+     * Sets the component's gradeable
+     * @param Gradeable $gradeable A non-null gradeable
+     */
+    private function setGradeable(Gradeable $gradeable)
+    {
+        if($gradeable === null) {
+            throw new \InvalidArgumentException('Gradeable Cannot be null!');
+        }
+        $this->gradeable = $gradeable;
+    }
+
+    /**
+     * Checks the point values to ensure consistency.  See `setPoints` docs for details
+     *
+     * @param $lower_clamp
+     * @param $default
+     * @param $max_value
+     * @param $upper_clamp
+     * @return array|null
+     */
+    private function validatePoints(&$lower_clamp, &$default, &$max_value, &$upper_clamp)
```

Comment:
Same as above. A function named "validate", I'd assume it would either return true/false if validation succeeded, not return either an array or null and then mutated the incoming variables. This goes back to above where the description of this function is "validates that parameters are numerics and casts them to floats". Those are two discrete statements and it's weird to see it all done within one function (and hence also why you're doing pass by value and returning things).

Given you probably want to give a status report at the end, I'd do make this function collect errors in an array, and if count($array) > 0, throw ValidationError, else do nothing. You then have roundPointValue which accepts a numeric string or float to handle the floatval for you.